compiler/src/iree/compiler/Dialect/HAL/Transforms/PreprocessExecutables.cpp

```diff
@@ -0,0 +1,321 @@
+                                           StringRef rawCommand) {
+  auto command = fixupArg(rawCommand);
+
+  // Setup IO redirects used to pass around the executable MLIR contents.
+  SmallString<32> stdinFile, stdoutFile, stderrFile;
+  int inputFd = 0;
+  llvm::sys::fs::createTemporaryFile("executable-preprocessor-stdin", "",
+                                     inputFd, stdinFile);
+  llvm::sys::fs::createTemporaryFile("executable-preprocessor-stdout", "",
+                                     stdoutFile);
+  llvm::sys::fs::createTemporaryFile("executable-preprocessor-stderr", "",
+                                     stderrFile);
+  llvm::FileRemover stdinRemover(stdinFile.c_str());
+  llvm::FileRemover stdoutRemover(stdoutFile.c_str());
+  llvm::FileRemover stderrRemover(stderrFile.c_str());
+  std::optional<StringRef> redirects[] = {
+      stdinFile.str(),
+      stdoutFile.str(),
+      stderrFile.str(),
+  };
+
+  // Serialize the executable contents.
+  // NOTE: this is currently being done in text format as it's easier to work
+  // with. We'll probably want to flip to binary or make it an option if we
+  // ever want to support versioning.
+  {
+    llvm::raw_fd_ostream inputStream(inputFd, /*shouldClose=*/true);
+    executableOp.print(inputStream,
+                       OpPrintingFlags().useLocalScope().enableDebugInfo());
+    inputStream << "\n";  // newline at end of file
+  }
+
+  // LLVM wants all the args split up to launch the command so we tokenize here.
+  // This is exactly how the LLVM command line parser does it with a macro
+  // switch.
+  llvm::BumpPtrAllocator scratchAllocator;
+  llvm::StringSaver stringSaver(scratchAllocator);
+  SmallVector<const char *> rawArgs;
+#ifdef _WIN32
+  auto Tokenize = llvm::cl::TokenizeWindowsCommandLine;
+#else
+  auto Tokenize = llvm::cl::TokenizeGNUCommandLine;
+#endif  // _WIN32
+  Tokenize(command, stringSaver, rawArgs, /*MarkEOLs=*/false);
+  SmallVector<StringRef> args;
+  for (auto rawArg : rawArgs) args.push_back(StringRef(rawArg));
+
+  // Try to find the tool either by absolute path or by looking it up in env.
+  auto tool = findTool(args[0].str());
+  if (tool.empty()) {
+    llvm::errs() << "ERROR: failed to find tool `" << args[0] << "` in PATH\n";
+    return failure();
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Launching hal.executable preprocessor: ";
+    for (auto arg : args) llvm::dbgs() << arg << " ";
+    llvm::dbgs() << " 1> " << stdoutFile.str() << " 2> " << stderrFile.str()
+                 << "\n";
+  });
+
+  // Launch the preprocessing tool. Note that this may fail for tons of reasons
+  // (bad program path, bad command line, bad system state, bad IO, bad
+  // preprocessor, etc).
+  std::string errorMessage;
+  int runResult = llvm::sys::ExecuteAndWait(
+      unescapeCommandLineComponent(tool), args, /*Env=*/std::nullopt,
+      /*Redirects=*/redirects,
+      /*SecondsToWait=*/0, /*MemoryLimit=*/0, /*ErrMsg=*/&errorMessage);
```

Comment:
with nullopt it inherits the calling process environment so that if the user sets a path or whatever it'll go through - if you override it here then (I believe) it stops passing them through and instead just passes what you give it